File: 13b.pl
#! /usr/bin/perl
use warnings;
use strict;
use feature qw{ say };
use experimental 'signatures';

use ARGV::OrDATA;
use List::Util qw{ any };
use Storable qw{ dclone };

sub parse($packet) {
    my @list;
    while ($packet =~ s/((?:,?[0-9]+)+|[\[\]])//) {
        my $token = $1;
        if ($token =~ /[0-9]/) {
            $token =~ s/^,+//;
            push @list, split /,/, $token;

        } elsif ('[' eq $token) {
            my ($parsed, $rest) = parse($packet);
            push @list, $parsed;
            $packet = $rest;

        } elsif (']' eq $token) {
            return \@list, $packet

        } else {
            die "Can't parse '$packet'";
        }
    }
    return @list
}

sub compare(@packets) {
    while (@{ $packets[1] }) {
        my $p1 = shift @{ $packets[0] };
        my $p2 = shift @{ $packets[1] };
        return 1 unless defined $p1;

        if (! ref $p1 && ! ref $p2) {
            return $p1 < $p2 ? 1 : -1 if $p1 != $p2;

        } elsif (! ref $p1 || ! ref $p2) {
            unshift @{ $packets[1] }, ref $p2 ? $p2 : [$p2];
            unshift @{ $packets[0] }, ref $p1 ? $p1 : [$p1];
        } else {
            my $cmp = compare($p1, $p2);
            return $cmp if defined $cmp;
        }
    }
    return -1 if @{ $packets[0] };

    return
}

my @packets = ([[2]], [[6]]);
local $/ = "";
while (<>) {
    push @packets, map { parse($_) } split;
}
@packets = sort { compare(dclone($b), dclone($a)) } @packets;
my $key = 1;
for my $i (0 .. $#packets) {
    $key *= $i + 1 if ref $packets[$i] && 1 == @{ $packets[$i] }
                   && ref $packets[$i][0] && 1 == @{ $packets[$i][0] }
                   && ! ref $packets[$i][0][0]
                   && any { $packets[$i][0][0] == $_} 2, 6;
}
say $key;

__DATA__
[1,1,3,1,1]
[1,1,5,1,1]

[[1],[2,3,4]]
[[1],4]

[9]
[[8,7,6]]

[[4,4],4,4]
[[4,4],4,4,4]

[7,7,7,7]
[7,7,7]

[]
[3]

[[[]]]
[[]]

[1,[2,[3,[4,[5,6,7]]]],8,9]
[1,[2,[3,[4,[5,6,0]]]],8,9]

File: 13b.py
#! /usr/bin/python3
import sys
import re
import copy
from functools import cmp_to_key

def _parse(tokens):
    struct = []
    while (tokens):
        if re.match(r'[0-9]', tokens[0]):
            struct += [int(tokens.pop(0))]

        elif '[' == tokens[0]:
            parsed, rest = _parse(tokens[1:])
            struct += [parsed]
            tokens = rest

        elif ']' == tokens[0]:
            return struct, tokens[1:]

        else:
            raise Exception(f"Can't parse {tokens[0]}")
    return struct[0]

def parse(packet):
    tokens = re.findall(r'([0-9]+|[\[\]])', line)
    return _parse(tokens)

def compare(*packets):
    while len(packets[1]) and not packets[1] is None:
        if not len(packets[0]):
            return 1

        p1 = packets[0].pop(0)
        p2 = packets[1].pop(0)

        if not type(p1) is list and not type(p2) is list:
            if p1 != p2:
                return 1 if p1 < p2 else -1

        elif not type(p1) is list or not type(p2) is list:
            packets[1].insert(0, p2 if type(p2) is list else [p2])
            packets[0].insert(0, p1 if type(p1) is list else [p1])

        else:
            comp = compare(p1, p2)
            if not comp is None:
                return comp
    if len(packets[0]):
        return -1

    return None

def sortcompare(*packets):
    return compare(copy.deepcopy(packets[1]), copy.deepcopy(packets[0]))

packets = [[[2]], [[6]]]
with open(sys.argv[1], 'r') as fh:
    for line in fh:
        if 1 == len(line):
            continue

        packets += [parse(line)]

key = 1
packets = sorted(packets, key=cmp_to_key(sortcompare))
for i in range(0, len(packets)):
    if 1 == len(packets[i]) and type(packets[i][0]) is list \
       and 1 == len(packets[i][0]) and not type(packets[i][0][0]) is list \
       and packets[i][0][0] in (2, 6):
        key *= i + 1

print(key)
